### PR TITLE
Embed amp-viewer into a Polymer element

### PR DESCRIPTION
**Pull request by kul3r4**
This PR moves the amp-viewer.js content into a proper Polymer element (version 2).
I haven't modified the README yet, please let me know if I can go ahead with that as well.

### CODE REVIEW

**Comment by frankiefu** [comment]
If your reason for changing amp-viewer to a Polymer element is to work with HTML imports (e.g. `<link rel="import" href="amp-viewer.html">` then you can simply just have a wrapper for it. For example, `amp-viewer.html`:
```html
<script src="amp-viewer.js"></script>
```
See `good-map.html` as an example: https://github.com/keanulee/good-map

To upgrade to CustomElements v1 which is what Polymer 2.x supports we can update `amp-viewer.js` to follow v1 spec: https://developers.google.com/web/fundamentals/web-components/customelements
See `good-map.js` as reference: https://github.com/keanulee/good-map/blob/master/good-map.js

cc @keanulee

**Comment by kul3r4** [comment]
@frankiefu Thanks a lot for the guidance!
I made some changes, please take a look.

File: bower.json [hunk 1]
```diff
@@ -11,6 +11,9 @@
     "type": "git",
     "url": "git://github.com/PolymerLabs/amp-viewer"
   },
+  "dependencies": {
+    "polymer": "polymer/polymer#^2.0.0"
```
Comment (TimvdLippe) [inline bower.json]: You are not using Polymer in this PR, so I think you can remove this again.

**Comment by kul3r4** [comment]
@TimvdLippe Thanks! just removed it.

File: amp-viewer.html [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<!--
+  @license
+  Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
```
Comment (valdrinkoshi) [inline amp-viewer.html]: (c) 2017

File: README.md [hunk 3]
```diff
@@ -8,7 +8,7 @@
 ```
 <custom-element-demo>
   <template>
-    <script src="amp-viewer.js"></script>
+    <script src="amp-viewer.html"></script>
```
Comment (valdrinkoshi) [inline README.md]: Revert this or make it a `<link rel=import>`

**Comment by kul3r4** [comment]
@valdrinkoshi PTAL

**Comment by valdrinkoshi** [comment]
@kul3r4 changes for the README.md and license approved 👌 Will review the functionality now (did the first review from the phone 📱)

File: amp-viewer.js [hunk 4]
```diff
@@ -8,116 +8,123 @@
...
-          this.src = this.getAttribute('src');
-        }
-      },
+    connectedCallback() {
 
-      set src(src) {
-        if (this._src === src) {
-          return;
-        }
+      if (!window.AMP_SHADOW) {
```
Comment (valdrinkoshi) [inline amp-viewer.js]: This was previously done in `createdCallback`, which corresponds to `constructor` in Custom Elements v1. This `if` should be moved up in the constructor not to change the timing

File: amp-viewer.js [hunk 5]
```diff
@@ -8,116 +8,123 @@
...
-        this._src = src;
-        this._clear();
-        if (!src) {
-          this.removeAttribute('src');
-        } else {
-          this.setAttribute('src', src);
-          this._loadDocument(src);
-        }
-      },
+      if (this.hasAttribute('src')) {
```
Comment (valdrinkoshi) [inline amp-viewer.js]: Custom Elements v1 has the [`observedAttributes`](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Custom_Elements#Observed_attributes) getter that allows you to tell which attributes you want to observe.
You should delete this check and just tell the browser you're interested on `src` attribute changes only, with:
```js
static get observedAttributes() {return ['src']; }

File: amp-viewer.js [hunk 6]
```diff
@@ -8,116 +8,123 @@
...
-        if (this._amp) {
-          this._amp.setVisibilityState(state);
-        }
-      },
+    get src() {
+      return this._src;
+    }
+
+    attributeChangedCallback(name, old, value) {
+      var desc = Object.getOwnPropertyDescriptor(this.__proto__,
```
Comment (valdrinkoshi) [inline amp-viewer.js]: observedAttributes ensures you get notified only of the changes on the attributes you care about, hence this whole method can be changed to just check for
```js
if (old !== value) this.src = value;

File: README.md [hunk 7]
```diff
@@ -8,7 +8,7 @@
 ```
 <custom-element-demo>
   <template>
-    <script src="amp-viewer.js"></script>
+    <link rel="import" href="amp-viewer.html" async>
```
Comment (valdrinkoshi) [inline README.md]: Question: any reason why to change this to an import? This comment is used by webcomponents.org to have a inline demo (see https://www.webcomponents.org/publish)
This element is not published yet on webcomponents.org, so I'd either not modify this line or completely remove code from line 8 to 15. @frankiefu WDYT?

Comment (frankiefu) [reply README.md]: Agree with @valdrinkoshi and let's remove it (line 8 to 15).

**Comment by kul3r4** [comment]
@valdrinkoshi Thanks for the comments, PTAL

**Comment by valdrinkoshi** [comment]
Nice job @kul3r4! I think we can keep the "main" as is in the `bower.json`, as `amp-viewer.html` is just a convenience file for whoever would like to import it through html imports.

Question for @frankiefu: the AMP script that is imported is using shadow dom v0 (see https://cdn.ampproject.org/shadow-v0.js), are you aware of any work done to update that to v1?

**Comment by frankiefu** [comment]
@valdrinkoshi Seems like they have just merged in a change that will make AMP to work with ShadowDOM v1: https://github.com/ampproject/amphtml/issues/9761

File: bower.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "amp-viewer",
-  "version": "0.0.1",
+  "version": "1.0.0",
```
Comment (valdrinkoshi) [inline bower.json]: @frankiefu is it fine to increment the version in this PR? or you'd prefer to do it in a separate PR?

File: demo/index.html [hunk 9]
```diff
@@ -33,7 +33,7 @@
 
   <amp-viewer id="viewer" src="sample-content/experience-daydream-today.html"></amp-viewer>
 
-  <script>
+  <script>
```
Comment (valdrinkoshi) [inline demo/index.html]: very nitty nit: problem with extra spaces here 👀

**Comment by kul3r4** [comment]
@valdrinkoshi PTAL

File: bower.json (hunk 8 above)
Comment (frankiefu) [reply bower.json]: I think this is fine.

**Comment by valdrinkoshi** [comment]
@kul3r4 few missing things (`fetchDocument -> _fetchDocument` and extra spaces from my last review), and then it should be good to go!

**Comment by kul3r4** [comment]
@valdrinkoshi I think I removed the spaces now, could you tell me again in which line you see them?

**Comment by kul3r4** [comment]
@valdrinkoshi Will these changed be merged?

**Comment by valdrinkoshi** [comment]
I leave that to the maintainers @frankiefu & @keanulee

**Comment by frankiefu** [comment]
@kul3r4 Thanks for the PR. @valdrinkoshi Thanks for reviewing the PR.
